crates/noirc_frontend/src/hir/type_check/expr.rs

```diff
@@ -698,18 +699,20 @@ pub fn comparator_operand_type_rules(
         (Error, _) | (_,Error) => Ok(Bool(Comptime::Yes(None))),
 
         // Special-case == and != for arrays
-        (Array(x_size, x_type), Array(y_size, y_type)) if matches!(op.kind, Equal | NotEqual) => {
-            x_type.unify(y_type, op.location.span, errors, || {
+        (Array{ elem, len }, Array { elem: y_elem, len: y_len }) if matches!(op.kind, Equal | NotEqual) => {
+            elem.unify(y_elem, op.location.span, errors, || {
                 TypeCheckError::Unstructured {
                     msg: format!("Cannot compare {} and {}, the array element types differ", lhs_type, rhs_type),
                     span: op.location.span,
                 }
             });
 
-            if x_size != y_size {
-                return Err(format!("Can only compare arrays of the same length. Here LHS is of length {}, and RHS is {} ", 
```

Comment:
This line is relevant to the original error: we should use `unify` so that we follow any type variables

Reply:
(these two comments were originally to point out the more relevant lines before I trimmed down this PR. I'll leave them here for clarity)